dali/pipeline/executor/executor.cc

```diff
@@ -296,21 +296,33 @@ void Executor<WorkspacePolicy, QueuePolicy>::RunHelper(OpNode &op_node, Workspac
   const auto &schema = spec.GetSchema();
   SmallVector<int, 16> empty_layout_in_idxs;
 
-  cudaStream_t prev_stage_stream = ws.has_stream() && ws.stream() == gpu_op_stream_
-    ? mixed_op_stream_ : gpu_op_stream_;
 
-  auto order = ws.has_stream() ? AccessOrder(ws.stream()) : AccessOrder::host();
-  auto set_order = [&](auto &output) {
-    // NOTE: the stage streams are synchronized by the executor
-    bool need_sync = output.order().stream() != prev_stage_stream;
+  auto ws_order = ws.has_stream() ? AccessOrder(ws.stream()) : AccessOrder::host();
+
+  // Try to infer from which stage the buffer comes from
+  auto order_to_stage_index = [&](AccessOrder order) {
+    if (!order || order == AccessOrder::host())
+      return 0;
+    else if (order == mixed_op_stream_)
+      return 1;
+    else if (order == gpu_op_stream_)
+      return 2;
+    else
+      return 3;  // foreign order, comes last
+  };
+
+  auto set_order = [&](auto &output, AccessOrder order) {
+    // Check if this stage (identified by ws_order) is already synchronized with the output's
```

Comment:
Apart from foreign order, can this situation actually appear with our internal orders? That would require, for example, for the GPU order to appear in the Mixed stage, which I don't think is easily achievable.

Reply:
No, it's not, but I think it should be - that is, setting the order in TensorLists that share storage should be observable in the others (but it's not that easy).